components/clp-mcp-server/clp_mcp_server/server/utils.py

```diff
@@ -9,6 +9,54 @@
 logger = logging.getLogger(__name__)
 
 
+def convert_date_string_to_epoch(date_string: str) -> int:
+    """
+    :param date_string: ISO 8601 formatted date string with millisecond precision
+        (YYYY-MM-DDTHH:mm:ss.fffZ).
+    :return: Unix epoch timestamp in milliseconds.
+    :raise: ValueError if `date_string` cannot be parsed or converted to a valid Unix epoch.
+    """
+    try:
+        cleaned_string = date_string.rstrip("Z")
+        dt = datetime.fromisoformat(cleaned_string).replace(tzinfo=timezone.utc)
+
+        # Convert to epoch with milliseconds precision.
+        epoch_seconds = dt.timestamp()
+        return int(epoch_seconds * 1000)
+
+    except (ValueError, AttributeError) as e:
+        err_msg = f"Invalid date string {date_string}: {e}."
+        raise ValueError(err_msg) from e
+
+
+def parse_timestamp_range(begin_timestamp: str, end_timestamp: str) -> tuple[int, int]:
+    """
+    :param begin_timestamp:
+    :param end_timestamp:
+    :return: begin and end timestamp in Unix epoch timestamp with milliseconds precision.
+    :raise: Propagates `convert_date_string_to_epoch`'s exceptions.
+    :raise: ValueError if `end_timestamp` is earlier than `begin_timestamp`.
+    :raise: ValueError if `begin_timestamp` or `end_timestamp` doesn't indicate UTC time.
```

Comment:
```suggestion
    :raise: ValueError if `end_timestamp` is earlier than `begin_timestamp`.
    :raise: ValueError if `begin_timestamp` or `end_timestamp` doesn't indicate UTC (missing
        trailing `Z`).
    :raise: Propagates `convert_date_string_to_epoch`'s exceptions.
```
A bit rewriting + reordering: "propagates" section should come after explicit exceptions.